File: leetcode-js/131-palindrome-partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    const result = [];
    dfs(s, 0, [], result);
    return result;
};

function dfs(s, index, solution, result) {
    if (index === s.length) {
        result.push(solution);
        return;
    }

    for (let i = index; i < s.length; i += 1) {
        if (isPal(s, index, i)) {
            dfs(s, i + 1, solution.concat(s.substring(index, i + 1)), result);
        }
    }
}

function isPal(s, left, right) {
    while (left < right) {
        if (s[left] !== s[right]) {
            return false;
        } else {
            left += 1;
            right -= 1;
        }
    }
    return true;
}

File: leetcode-java/131-palindrome-partitioning.java
public class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        dfs(s, res, new ArrayList<String>(), 0);
        return res;
    }

    void dfs(String s, List<List<String>> res, List<String> cur, int index) {
        if (index == s.length()) {
            res.add(new ArrayList<>(cur));
            return;
        }

        for (int i = index; i < s.length(); i++) {
            if (isPal(s, index, i)) {
                cur.add(s.substring(index, i + 1));
                dfs(s, res, cur, i + 1);
                cur.remove(cur.size() - 1);
            }
        }
    }

    boolean isPal(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left) == s.charAt(right)) {
                left++;
                right--;
            } else {
                return false;
            }
        }

        return true;
    }
}
